src/unstated.js

```diff
@@ -13,8 +13,11 @@ export class Container<State: {}> {
     this._listeners = [];
   }
 
-  setState(state: $Shape<State>) {
-    this.state = Object.assign({}, this.state, state);
+  setState(state: $Shape<State> | (State => $Shape<State>)) {
+    this.state =
+      typeof state === 'function'
+        ? state(this.state)
```

Comment:
Shouldn't this still be merged back into the existing state?

Reply:
That's a good point. In the example it would just allocate an unnecessary empty object and key-iteration round (with Object.assign), as `lodash/fp` / `ramda` take care of immutability.

Reply:
Hello, I'd prefer it to stay that way so it reflects React's default behaviour.
Thanks again @jamiebuilds for this excellent project.

Reply:
How about this?